python/lsst/analysis/tools/actions/keyedData/calcDistances.py

```diff
@@ -81,61 +89,139 @@ def __call__(self, data: KeyedData, **kwargs) -> KeyedData:
-        idx, idxCatalog, d2d, d3d = catalog.search_around_sky(catalog, annulus[1])
-        inAnnulus = d2d > annulus[0]
-        idx = idx[inAnnulus]
-        idxCatalog = idxCatalog[inAnnulus]
-
-        rmsDistances = []
-        sepResiduals = []
-        for id in range(len(meanRa)):
-            match_inds = idx == id
-            match_ids = idxCatalog[match_inds & (idxCatalog != id)]
-            if match_ids.sum() == 0:
-                continue
-
-            object_srcs = df.loc[df["groupKey"] == meanRa.index[id]]
-
-            object_visits = object_srcs["visit"].to_numpy()
-            object_ras = (object_srcs["coord_ra"].to_numpy() * u.degree).to(u.radian).value
-            object_decs = (object_srcs["coord_dec"].to_numpy() * u.degree).to(u.radian).value
-            if len(object_srcs) <= 1:
-                continue
-            object_srcs = object_srcs.set_index("visit")
-            object_srcs.sort_index(inplace=True)
-
-            for id2 in match_ids:
-                match_srcs = df.loc[df["groupKey"] == meanRa.index[id2]]
-                match_visits = match_srcs["visit"].to_numpy()
-                match_ras = (match_srcs["coord_ra"].to_numpy() * u.degree).to(u.radian).value
-                match_decs = (match_srcs["coord_dec"].to_numpy() * u.degree).to(u.radian).value
-
-                separations = matchVisitComputeDistance(
-                    object_visits, object_ras, object_decs, match_visits, match_ras, match_decs
-                )
-
-                if len(separations) > 1:
-                    rmsDist = np.std(separations, ddof=1)
-                    rmsDistances.append(rmsDist)
-                if len(separations) > 2:
-                    sepResiduals.append(separations - np.median(separations))
+        # Compute the mean from the ragged array of pairs by
+        # using np.add.at to sum numerator and denominator.
+        sepMean = np.zeros(len(i1))
+        nSep = np.zeros_like(sepMean, dtype=np.int32)
+        np.add.at(sepMean, matchedPairInd, separations)
+        np.add.at(nSep, matchedPairInd, 1)
+        good = nSep > 1
+        sepMean[good] /= nSep[good]
+        sepMean[~good] = np.nan
+
+        # There are no good pairs, so return the default.
+        if good.sum() == 0:
+            return distanceParams
+
+        # Compute the stdev with sqrt(sum((sep - mean(sep))**2.)/(nsep - 1))
+        sepStd = np.zeros_like(sepMean)
+        np.add.at(
+            sepStd,
+            matchedPairInd,
+            (separations - sepMean[matchedPairInd]) ** 2.0,
+        )
+        sepStd[good] = np.sqrt(sepStd[good] / (nSep[good] - 1))
+        rmsDistances = sepStd[good]
+
+        # Need sepResiduals, but only when nSep is > 2.
+        bad2 = nSep <= 2
+        sepMean[bad2] = np.nan
+        sepResiduals = separations - sepMean[matchedPairInd]
+        sepResiduals = sepResiduals[np.isfinite(sepResiduals)]
 
         if len(rmsDistances) == 0:
```

Comment:
I don't think you need this anymore, since you already returned if `good.sum() == 0`